### PR TITLE
Expense module: UI and Integration

### PR DESCRIPTION
**Pull request by Shruti-Apte**
### What 
1. UI for all pages.
2. Edit and Delete methods added expense_controller
3. API Integration
4. Mobile responsive pages.

Preview:
Screenshot 2024-01-09 at 4 23 12 PM

Screenshot 2024-01-09 at 4 22 42 PM

Screenshot 2024-01-09 at 4 22 50 PM

Screenshot 2024-01-09 at 4 23 26 PM

Screenshot 2024-01-09 at 4 23 36 PM

### What's Remaining
- Download expense functionality.
- Expense summary API and integration
- PieChart
- Client list API is not present

### CODE REVIEW

File: app/controllers/internal_api/v1/expenses_controller.rb [hunk 1]
```diff
@@ -27,6 +27,22 @@ def show
     render :show, locals: { expense: Expense::ShowPresenter.new(@expense).process }
   end
 
+  def update
+    authorize set_expense
+
+    @expense.update(expense_params)
+
```
Comment (apoorv1316) [inline app/controllers/internal_api/v1/expenses_controller.rb:34]: Use the bang method ( update!)

File: app/controllers/internal_api/v1/expenses_controller.rb [hunk 2]
```diff
@@ -27,6 +27,22 @@ def show
...
+    render json: { notice: "Expense updated successfully" }, status: :ok
```
Comment (apoorv1316) [inline app/controllers/internal_api/v1/expenses_controller.rb:35]: Use locales.

Example:     `render json: { notice: I18n.t("invitation.update.success.message") }, status: :ok`

File: app/controllers/internal_api/v1/expenses_controller.rb [hunk 3]
```diff
@@ -27,6 +27,22 @@ def show
...
+  end
+
+  def destroy
+    authorize set_expense
+
+    @expense.destroy
```
Comment (apoorv1316) [inline app/controllers/internal_api/v1/expenses_controller.rb:41]: Use bang method for destroy

File: app/controllers/internal_api/v1/expenses_controller.rb [hunk 4]
```diff
@@ -27,6 +27,22 @@ def show
...
+
+    render json: { notice: "Expense deleted successfully" }, status: :ok
```
Comment (apoorv1316) [inline app/controllers/internal_api/v1/expenses_controller.rb:43]: Use locales

**Review by apoorv1316 — CHANGES_REQUESTED** [review]

File: package.json [hunk 5]
```diff
@@ -27,6 +27,7 @@
     "babel-loader": "^9.1.2",
     "babel-plugin-js-logger": "^1.0.17",
     "babel-plugin-transform-react-remove-prop-types": "^0.4.24",
+    "chart.js": "^4.1.2",
```
Comment (JijoBose) [inline package.json:30]: please push the changes added to `yarn.lock`

**Review by JijoBose — CHANGES_REQUESTED** [review]
Screenshot 2024-01-12 at 1 38 08 PM

**Comment by JijoBose** [comment]
@Shruti-Apte Filter button has no functionality, 
Should we remove it if it's not part of this Ticket?

**Comment by apoorv1316** [comment]
@Shruti-Apte Please work on changes discussed on call

File: app/controllers/internal_api/v1/expenses_controller.rb [hunk 6]
```diff
@@ -27,6 +27,22 @@ def show
     render :show, locals: { expense: Expense::ShowPresenter.new(@expense).process }
   end
 
+  def update
+    authorize set_expense
+
```
Comment (apoorv1316) [inline app/controllers/internal_api/v1/expenses_controller.rb:32]: this is incorrect. We have to authorize Expense here and not the private method

File: app/controllers/internal_api/v1/expenses_controller.rb [hunk 7]
```diff
@@ -27,6 +27,22 @@ def show
...
+    @expense.update!(expense_params)
+
+    render json: { notice: I18n.t("expenses.update") }, status: :ok
+  end
+
+  def destroy
+    authorize set_expense
```
Comment (apoorv1316) [inline app/controllers/internal_api/v1/expenses_controller.rb:39]: ditto

**Review by apoorv1316 — CHANGES_REQUESTED** [review]
@Shruti-Apte Please fix the requested changes

File: app/javascript/src/components/Expenses/Details/index.tsx [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+        />
+      )}
+      {showDeleteExpenseModal && (
+        <DeleteExpenseModal
+          handleDeleteExpense={handleDeleteExpense}
+          setShowDeleteExpenseModal={setShowDeleteExpenseModal}
+          showDeleteExpenseModal={showDeleteExpenseModal}
+        />
+      )}
+      {showEditExpenseModal && !isDesktop && (
```
Comment (nisusam) [inline app/javascript/src/components/Expenses/Details/index.tsx:110]: Move this above `DeleteExpenseModal`

File: app/javascript/src/components/Expenses/List/Container/Table/MoreOptions.tsx [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+  return isDesktop ? (
+    <div
+      className="absolute bottom-16 right-0 hidden items-center justify-between rounded-xl border-2 border-miru-gray-200 bg-white lg:w-28 lg:p-2 lg:group-hover:flex xl:w-40 xl:p-3"
+      onClick={e => e.stopPropagation()}
+    >
+      <Tooltip content="Download">
+        <Button
+          disabled
+          style="ternary"
+          onClick={e => {
```
Comment (nisusam) [inline app/javascript/src/components/Expenses/List/Container/Table/MoreOptions.tsx:24]: If the button is disabled, do we need to add `onClick`?

File: app/javascript/src/components/Expenses/Modals/ExpenseForm.tsx [hunk 10]
```diff
@@ -0,0 +1,332 @@
...
+          size={16}
+          weight="bold"
+        />
+      </div>
+      <div className="ml-4 mr-2 flex w-full flex-col items-start truncate">
+        <span className="text-sm font-medium">{receipt.name}</span>
+        <div className="flex items-center text-xs font-medium text-miru-dark-purple-400">
+          <span>PDF</span>
+          <div className="mx-2 h-1 w-1 rounded-xl bg-miru-dark-purple-200" />
+          <span>{Math.ceil(receipt.size / 1000)}kb</span>
```
Comment (nisusam) [inline app/javascript/src/components/Expenses/Modals/ExpenseForm.tsx:156]: This should be 1024.

**Review by nisusam — CHANGES_REQUESTED** [review]

File: app/javascript/src/components/Expenses/List/Container/Table/TableRow.tsx [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+          <div className="flex h-full flex-col items-start">
+            <span className="font-semibold lg:font-medium">{expenseType}</span>
+            <dl className="truncate text-left text-sm font-medium leading-5 text-miru-dark-purple-400 lg:hidden">
+              <dt className="mt-1">{date}</dt>
+            </dl>
+          </div>
+        </td>
+        <td className="relative w-1/5 whitespace-nowrap py-4 pl-4 text-right text-sm lg:table-cell lg:w-2/12 lg:pr-2 lg:text-xl lg:font-bold">
+          {currencyFormat(currency, amount)}
+          {isDesktop && (
```
[the remark this answers is not in the document]
Comment (Shruti-Apte) [reply app/javascript/src/components/Expenses/List/Container/Table/TableRow.tsx:76]: No, It cannot be refactored into Ternary

**Review by apoorv1316 — APPROVED** [review]
Some requirements are yet to be implemented. Let's do it in upcoming PR.

**Review by nisusam — APPROVED** [review]

**Pull request merged**